File: gcj/gcj2017/qual/c/C.java
import static java.lang.Math.*;
import static java.util.Arrays.*;
import java.io.*;
import java.util.*;

public class C
{
    Scanner sc = new Scanner(System.in);

    Long[] split_two(Long n) {
        Long mn;
        Long mx;
        mn = mx = (n - 1) / 2;
        if (mn + mx < n - 1) {
            mx += 1L;
        }
        Long ret[] = {mx, mn};
        return ret;
    }

    void add_num(PriorityQueue<Long> queue, Map<Long, Long> map,
                 Long n, Long num) {
        if (map.get(n) == null) {
            queue.add(-n);
            map.put(n, num);
        }
        else {
            map.put(n, map.get(n) + num);
        }
    }

    void solve() {
        PriorityQueue<Long> queue = new PriorityQueue<Long>(10); 
        Map<Long, Long> map = new HashMap<Long, Long>();
        Long N = sc.nextLong();
        Long K = sc.nextLong();
        queue.add(-N);
        map.put(N, 1L);
        
        while (K > 0) {
            Long n = -queue.remove();
            Long num = map.get(n);
            map.remove(n);
            if (num >= K) {
                Long[] mxmn = split_two(n);
                System.out.printf("%d %d", mxmn[0], mxmn[1]);
                return;
            }
            else {
                K -= num;
                Long[] mxmn = split_two(n);
                Long mx = mxmn[0];
                Long mn = mxmn[1];
                if (mx == mn && mx != 0) {
                    add_num(queue, map, mx, num * 2);
                }
                else {
                    if (mn != 0) {
                        add_num(queue, map, mn, num);
                    }
                    add_num(queue, map, mx, num);
                }
            }
        }
   }

    void run() {
        int T = sc.nextInt();
        for (int t = 1; t <= T; ++t) {
            System.out.printf("Case #%d: ", t);
            solve();
            System.out.printf("\n");
            System.out.flush();
        }
    }

    public static void main(String[] args) {
        new C().run();
    }
}

File: gcj/gcj2017/qual/c/c.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import collections
import queue

def simulate(N, K):
    arr = [1] + [0] * N + [1]

    for k in range(K):
        max_pos = -1
        max_d_l = -1
        max_d_r = -1
        for i in range(1, N + 1):
            if arr[i] == 1:
                continue
            dist_r = arr.index(1, i) - i - 1
            rev_i = len(arr) - i - 1
            dist_l = arr[::-1].index(1, rev_i) - rev_i - 1
#            print("dist_l, dist_r = ", dist_l, dist_r)
            if min(dist_l, dist_r) > min(max_d_l, max_d_r) or \
            min(dist_l, dist_r) == min(max_d_l, max_d_r) and \
            max(dist_l, dist_r) > max(max_d_l, max_d_r):
                max_d_l = dist_l
                max_d_r = dist_r
                max_pos = i
        arr[max_pos] = 1

    return tuple(reversed(sorted([max_d_l, max_d_r])))


def split_two(n):
    mn = mx = (n - 1) // 2
    if mn + mx < n - 1:
        mx += 1
    return (mx, mn)
    
def add_num(q, c, n, num):
    if n not in c:
        q.put(-n)
    c[n] += num

def fast(N, K):
    q = queue.PriorityQueue()
    q.put(-N)
    c = collections.Counter()
    c[N] = 1
    while K > 0:
        n = -q.get()
        num = c[n]
        del c[n]
        if num >= K:
            return split_two(n)
        else:
            K -= num
            mx, mn = split_two(n)
            if mx == mn and mx != 0:
                add_num(q, c, mx, num * 2)
            else:
                if mn != 0:
                    add_num(q, c, mn, num)
                add_num(q, c, mx, num)


# for n in range(1, 1001):
#     for k in range(1, n + 1):
#         ans1 = simulate(n, k)
#         ans2 = fast(n, k)
#         if ans1 != ans2:
#             print("error! n, k, sim, fast = ", n, k, ans1, ans2)
#             sys.exit(0)
#             break
#         else:
#             pass

def solve():
    N, K = map(int, input().split())
    print(" ".join(str(x) for x in fast(N, K)))

def main():
    T = int(input())
    for t in range(T):
        print ("Case #" + str(t+1) + ": ", end="")
        solve()

main()
